<a href="https://colab.research.google.com/github/davidlfc/Imersao_IA_ALURA_GOOGLE/blob/main/C%C3%B3pia_de_EP1_GPS_Newton_alunos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MAP3121 - Exercício programa 1 - 2025 - Turmas 4 a 11

## GPS e o Método de Newton

## O sistema de posicionamento global

<p align = "justify">O Sistema de Posicionamento Global (GPS, sigla do nome em inglês) é um
sistema de navegação formado por uma rede de pelo menos 24 satélites orbitando a Terra e
transmitindo sinais para os receptores GPS, que usam triangulação para calcular
a localização do usuário. Na descrição a seguir, o sistema de coordenadas tem
sua origem $O$ no centro da Terra, o eixo $Oz$ positivo aponta na direção do Polo
Norte, o plano $Oxy$ é o plano do equador com o eixo $Ox$ positivo cortando o
meridiano de Greenwich e o eixo $Oy$ positivo corta o meridiano de longitude
$90^{\circ}$E.</p>

<p align = "justify">Um dado satélite $i$ transmite a sua posição atual $(x_i , y_i , z_i)$ e o instante $te_i$
de emissão do sinal. O receptor GPS grava esta informação juntamente com
o instante $tr_i$ da recepção do sinal. Entretanto, o relógio do receptor é menos
preciso do que o relógio do satélite, havendo um erro de sincronia $T$ . O instante
correto da recepção é $tr_i - T$ . Se $T > 0$ o relógio do receptor está adiantado e
se $T < 0$ ele está atrasado.</p>

<p align = "justify">Portanto, o tempo que o sinal leva do satélite ao receptor é $t_i - T$ , onde
$t_i = tr_i - te_i$ é o lapso de tempo aparente, e o receptor deve estar na superfı́cie
da esfera de raio $c(t_i - T)$ e centro $(x_i , y_i , z_i)$, onde $c = 299792458$ m/s é a
velocidade da luz no vácuo. As coordenadas $(x, y, z)$ do receptor devem então
satisfazer a relação</p>

\begin{equation*}
  (x - x_i)^2 + (y - y_i)^2 + (z - z_i)^2 = (w_i - w)^2
\end{equation*}

<p align = "justify">em que $w_i = ct_i$ e $w = cT$ (ao multiplicarmos por c, ficamos apenas com dimensões
espaciais).</p>

<p align = "justify">Em qualquer instante, dados de 5 a 8 satélites são obtidos pelo receptor em
qualquer ponto da terra. Então, assumindo que temos dados de $n > 4$ satélites,
as 4 incógnitas $(x, y, z, w)$ devem satisfazer as equações</p>

\begin{equation}
  r_i(x,y,z,w) = (x - x_i)^2 + (y - y_i)^2 + (z - z_i)^2 - (w_i - w)^2 = 0, \tag{1}
\end{equation}

<p align = "justify">$1 \le i \le n$, que formam um sistema não-linear sobredeterminado. Como as
medidas estão sujeitas a erros, este sistema em geral não tem solução (se não
houvesse erros, bastaria resolver 4 das n equações).</p>

## Mínimos quadrados não-lineares

Uma maneira de se resolver o problema acima é tratá-lo como um problema
de mı́nimos quadrados, procurando-se **minimizar a soma dos quadrados dos resı́duos** $r_i$ , ou seja, determinando-se $\bar x$, $\bar y$, $\bar z$ e $\bar w$ que minimizam

$$
  g(x,y,z,w) = \sum_{i=1}^n r_i(x,y,z,w)^2.
$$

Uma condição necessária que um ponto de mínimo deve satisfazer é

\begin{equation}
  \nabla g(x, y, z, w) = 0 \tag{2}
\end{equation}

<p align = "justify">em que $\nabla g$ é o gradiente de $g$ em relação a $x$, $y$, $z$ e $w$. A condição acima é um
sistema não-linear de 4 equações e 4 incógnitas, que será resolvido usando o
método de Newton.</p>

## Mínimos quadrados lineares

<p align = "justify">Suponha que as equações $(1)$ são exatas. Se subtrairmos a equação $n$ da
equação $i$, os termos quadráticos cancelam e portanto das relações</p>

\begin{equation}
  r_i(x,y,z,w) - r_n(x,y,z,w) = 0,\quad 1 \le i \le n-1 \tag{3}
\end{equation}

<p align = "justify">obtemos um sistema linear sobredeterminado para as incógnitas que pode ser
resolvido no sentido de mı́nimos quadrados como foi visto
em aula.</p>

<p align = "justify">Se $A$ denota a matriz do sistema linear, $\mathbf{x}$ o vetor incógnita e $\mathbf{b}$ o lado
direito do sistema linear, a solução de $A\mathbf{x} = \mathbf{b}$ no sentido de mı́nimos quadrados
é a solução do sistema normal $A^TA\mathbf{x} = A^T\mathbf{b}$. Esta solução pode ser usada
como uma aproximação em si para a localização usando GPS, e também como
aproximação inicial para a resolução de $(2)$ usando o método de Newton.</p>

## ❗Exercício 1

<p align = "justify">Escreva explicitamente o sistema não-linear $(2)$ e o sistema linear sobredeterminado $(3)$. Você não precisa entregar as respostas. Conhecê-las é importante para as implementações.</p>

In [19]:
# Sistema não-linear (gradiente de g)
from sympy import symbols, Eq, diff, Function, summation, simplify

# Variáveis simbólicas
x, y, z, w = symbols('x y z w')
xi, yi, zi, wi = symbols('x_i y_i z_i w_i')
n = symbols('n', integer=True)

# Residuo simbólico genérico
r = (x - xi)**2 + (y - yi)**2 + (z - zi)**2 - (wi - w)**2

# Função objetivo g = soma dos resíduos ao quadrado
g_i = r**2

# Derivadas parciais (gradiente de g)
dg_dx = diff(g_i, x)
dg_dy = diff(g_i, y)
dg_dz = diff(g_i, z)
dg_dw = diff(g_i, w)

# sistema não-linear ∇g = 0
print("Sistema não-linear (∇g = 0):")
display(Eq(diff(g_i, x), 0))
display(Eq(diff(g_i, y), 0))
display(Eq(diff(g_i, z), 0))
display(Eq(diff(g_i, w), 0))


Sistema não-linear (∇g = 0):


Eq((4*x - 4*x_i)*(-(-w + w_i)**2 + (x - x_i)**2 + (y - y_i)**2 + (z - z_i)**2), 0)

Eq((4*y - 4*y_i)*(-(-w + w_i)**2 + (x - x_i)**2 + (y - y_i)**2 + (z - z_i)**2), 0)

Eq((4*z - 4*z_i)*(-(-w + w_i)**2 + (x - x_i)**2 + (y - y_i)**2 + (z - z_i)**2), 0)

Eq((-4*w + 4*w_i)*(-(-w + w_i)**2 + (x - x_i)**2 + (y - y_i)**2 + (z - z_i)**2), 0)

In [22]:
# Sistema linear sobredeterminado (diferença entre r_i e r_n)
xi, yi, zi, wi = symbols('x_i y_i z_i w_i')
xn, yn, zn, wn = symbols('x_n y_n z_n w_n')

# r_i e r_n
ri = (x - xi)**2 + (y - yi)**2 + (z - zi)**2 - (wi - w)**2
rn = (x - xn)**2 + (y - yn)**2 + (z - zn)**2 - (wn - w)**2

# Subtrair r_i - r_n
dif = simplify(ri - rn)

print("Sistema linear sobredeterminado (r_i - r_n = 0):")
display(Eq(ri - rn, 0))
display(Eq(dif, 0))


Sistema linear sobredeterminado (r_i - r_n = 0):


Eq(-(-w + w_i)**2 + (-w + w_n)**2 + (x - x_i)**2 - (x - x_n)**2 + (y - y_i)**2 - (y - y_n)**2 + (z - z_i)**2 - (z - z_n)**2, 0)

Eq(-(w - w_i)**2 + (w - w_n)**2 + (x - x_i)**2 - (x - x_n)**2 + (y - y_i)**2 - (y - y_n)**2 + (z - z_i)**2 - (z - z_n)**2, 0)

## Método de Newton para sistemas não lineares

<p align = "justify">Vamos descrever o método de Newton para determinação de raı́zes de funções $F(x)$ de $\mathbb{R}^n$ em $\mathbb{R}^n$. Como no caso unidimensional, parte-se de uma aproximação inicial $x^{0}$ para o valor $\bar{x}\in\mathbb{R}^n$ tal que $F(\bar{x})=0$ e calcula-se a sequência

$$
  x^{(k+1)} = x^{(k)} - J_F^{-1}(x^{(k)})F(x^{(k)})
$$
em que $J_F(x)$ é a matriz Jacobiana de $F$avaliada no ponto $x$:

$$
  J_F(x) = \left[\begin{array}{cccc}
  \frac{\partial F_1(x)}{\partial x_1} & \frac{\partial F_1(x)}{\partial x_2} & \cdots & \frac{\partial F_1(x)}{\partial x_n}\\
  \frac{\partial F_2(x)}{\partial x_1} & \frac{\partial F_2(x)}{\partial x_2} & \cdots & \frac{\partial F_2(x)}{\partial x_n}\\
  \vdots & \vdots & \cdots & \vdots\\
  \frac{\partial F_n(x)}{\partial x_1} & \frac{\partial F_n(x)}{\partial x_2} & \cdots & \frac{\partial F_n(x)}{\partial x_n}
  \end{array}\right],
$$

onde $F_1(x)$, $F_2(x)$, $\cdots$, $F_n(x)$ são as componentes de $F(x)$.</p>

Se reescrevermos as iterações do Método de Newton como

$$
  J_F(x^{(k)})(x^{(k+1)}-x^{(k)}) = -F(x^{(k)}),
$$

<p align = "justify">é possível perceber que não é necessário inverter a matriz Jacobiana de $F$ a cada passo, que seria muito custoso. A cada passo do Método de Newton, resolve-se o sistema linear</p>

$$
  J_F(x^{(k)})\cdot\mathbf{c} = -F(x^{(k)})
$$

e calcula-se a nova aproximação como

$$
  x^{(k+1)} = x^{(k)} + \mathbf{c}
$$

<p align = "justify">Quando F é de classe
$C^2$ e a matriz Jacobiana é não singular na raiz de $F$ (ou seja, tem determinante
não nulo), pode-se mostrar que o método de Newton converge quadraticamente
para a raiz de $F$ , desde que $x^{(0)}$ seja escolhido suficientemente próximo da raiz.</p>

## ❗Exercício 2

Para $F(x,y,z,w) = \nabla g(x,y,z,w)$, qual é a expressão para $J_F(x,y,z,w)$?

In [21]:
import sympy as sp
from IPython.display import display

# Variáveis
x, y, z, w = sp.symbols('x y z w')
xi, yi, zi, wi = sp.symbols('x_i y_i z_i w_i')

# Residuo r_i
r = (x - xi)**2 + (y - yi)**2 + (z - zi)**2 - (wi - w)**2

# g_i = r_i^2
g_i = r**2

# F = gradiente de g_i
F = sp.Matrix([
    sp.diff(g_i, x),
    sp.diff(g_i, y),
    sp.diff(g_i, z),
    sp.diff(g_i, w)
])

# Jacobiana de F (Hessiana de g_i)
JF = F.jacobian([x, y, z, w])

# Exibir a matriz Jacobiana
print("Matriz Jacobiana J_F(x, y, z, w) = Hessiana de g:")
display(JF)


Matriz Jacobiana J_F(x, y, z, w) = Hessiana de g:


Matrix([
[-4*(-w + w_i)**2 + 4*(x - x_i)**2 + (2*x - 2*x_i)*(4*x - 4*x_i) + 4*(y - y_i)**2 + 4*(z - z_i)**2,                                                                       (4*x - 4*x_i)*(2*y - 2*y_i),                                                                       (4*x - 4*x_i)*(2*z - 2*z_i),                                                                       (-2*w + 2*w_i)*(4*x - 4*x_i)],
[                                                                      (2*x - 2*x_i)*(4*y - 4*y_i), -4*(-w + w_i)**2 + 4*(x - x_i)**2 + 4*(y - y_i)**2 + (2*y - 2*y_i)*(4*y - 4*y_i) + 4*(z - z_i)**2,                                                                       (4*y - 4*y_i)*(2*z - 2*z_i),                                                                       (-2*w + 2*w_i)*(4*y - 4*y_i)],
[                                                                      (2*x - 2*x_i)*(4*z - 4*z_i),                                                                       (2*y - 2*y_i)*(4*z - 4*

# Tarefa

<p align = "justify">Você deve implementar um algoritmo para resolver um sistema não linear com $m$ equações e $m$ incógnitas.
Várias funções são necessárias para esta implementação. Comecemos por carregar o numpy. É o $\mathbf{único}$ módulo do Python que você irá usar.</p>

In [1]:
import numpy as np

def newton_system(F, JF, x0, tol=1e-8, max_iter=100):
    """
    Resolve F(x) = 0 usando o Método de Newton para sistemas não lineares.

    Parâmetros:
    - F: função que recebe um vetor x e retorna um vetor F(x) ∈ R^m
    - JF: função que recebe um vetor x e retorna a matriz Jacobiana JF(x) ∈ R^{m x m}
    - x0: aproximação inicial (numpy array)
    - tol: tolerância para critério de parada (norma do incremento)
    - max_iter: número máximo de iterações

    Retorno:
    - vetor solução x ∈ R^m
    """
    x = x0.astype(float)
    for k in range(max_iter):
        Fx = F(x)
        J = JF(x)
        try:
            delta = np.linalg.solve(J, -Fx)
        except np.linalg.LinAlgError:
            raise ValueError("Matriz Jacobiana singular na iteração", k)
        x = x + delta
        if np.linalg.norm(delta, ord=2) < tol:
            print(f"Convergência em {k+1} iterações.")
            return x
    raise ValueError("Método de Newton não convergiu após", max_iter, "iterações.")


## Resolução de sistemas lineares

Implemente a decomposição $LU$ de uma matriz $A$, com condensação pivotal, de acordo com o algoritmo

Dados $n$ e uma matrz $A$ (n\times n)$ temos:

&ensp;  $\bullet$ Para $k$ de $1$ a $n$ faça

&emsp; &emsp; - para $i$ de $k$ a $n$ faça $$a_{ik}=a_{ik}-\sum_{j=1}^{k-1}a_{ij}*a_{jk}$$

&emsp; &emsp; - Determine $l\ge k$ tal que $\vert a_{lk}\vert=\max_{k\le i\le n}\vert a_{ik}\vert$

&emsp; &emsp; - defina $p(k)=l$

&emsp; &emsp; - se $k\ne p(k)$ troque as linhas $k$ e $p(k)$ da matriz $A$

&emsp; &emsp; - para $j$ de $k+1$ a $n$ faça \begin{align} &a_{kj}=a_{kj}-\sum_{i=1}^{k-1}a_{ki}*a_{ij}\\ &a_{jk} = a_{jk}/a_{kk}\end{align}

**Observações:**

1. Ao final do algoritmo a matriz $L$ tem seus valores abaixo da diagonal
principal armazenados nas posições correspondentes de A (lembre-se que
a diagonal de L é composta de $1'$s).

2. A matriz $U$ tem seus valores da diagonal principal e acima desta armazenados nas posições correspondentes de $A$.

3. A decomposição $LU$ calculada corresponde à matriz $A$ permutada. As permutações realizadas estão armazenadas no vetor $p$ definido no algoritmo.

4. Lembre-se que ao final do algoritmo a matriz $A$ foi modificada. Caso esta
ainda seja necessária, uma cópia sua deve ser anteriormente salva.

5. Somatórios de $1$ a $0$ e loops de $n + 1$ a $n$ devem ser entendidos como vazios.

In [26]:
import numpy as np

def lu_dcmp(A):
    """
    Calcula a decomposição LU da matriz A com condensação pivotal,
    armazenando L (sem diagonal) abaixo da diagonal e U na diagonal e acima.

    Parâmetros:
    - A: matriz quadrada n x n (numpy array)

    Retorna:
    - A_cp: matriz LU modificada com L e U embutidas
    - p: vetor de permutação das linhas
    """

    A_cp = np.copy(A).astype(float)  # Garantir precisão float
    n = A_cp.shape[0]
    p = np.arange(n)  # vetor de permutação de linhas

    for k in range(n):
        # Etapa 1: atualizar a coluna k da linha k até n
        for i in range(k, n):
            soma = sum(A_cp[i, j] * A_cp[j, k] for j in range(k))
            A_cp[i, k] = A_cp[i, k] - soma

        # Etapa 2: escolher pivô máximo na coluna k
        l = np.argmax(np.abs(A_cp[k:n, k])) + k  # índice global
        p[k] = l

        # Etapa 3: trocar linhas k e l, se necessário
        if k != l:
            A_cp[[k, l], :] = A_cp[[l, k], :]

        # Etapa 4: atualizar linha k e os elementos abaixo de k
        for j in range(k+1, n):
            soma = sum(A_cp[k, i] * A_cp[i, j] for i in range(k))
            A_cp[k, j] = A_cp[k, j] - soma
            A_cp[j, k] = A_cp[j, k] / A_cp[k, k]

    return A_cp, p


Uma vez obtida a decomposição $LU$ da matriz $A$, calcule a solução do sistema linear $A\mathbf{x} = \mathbf{b}$.

In [27]:
import numpy as np

def lu_solve(LU, p, b):
    """
    Resolve Ax = b usando a decomposição LU com pivotamento armazenada em LU.

    Entradas:
    - LU: matriz que contém L e U embutidas
    - p: vetor de permutação de linhas
    - b: vetor do lado direito (n-dimensional)

    Saída:
    - x: vetor solução
    """
    n = len(b)
    y = np.zeros(n)
    x = np.zeros(n)

    # Aplicar permutação no vetor b: Pb
    b_perm = b.copy().astype(float)
    for k in range(n):
        if p[k] != k:
            b_perm[[k, p[k]]] = b_perm[[p[k], k]]

    # Substituição direta: resolver L y = Pb
    for i in range(n):
        y[i] = b_perm[i] - sum(LU[i, j] * y[j] for j in range(i))

    # Substituição retroativa: resolver U x = y
    for i in reversed(range(n)):
        x[i] = (y[i] - sum(LU[i, j] * x[j] for j in range(i+1, n))) / LU[i, i]

    return x


Com estas implementações, podemos usar o Método de Newton para a resolução de sistemas não lineares.

In [8]:
def newton_sitemas(F, JF, x0, tol, max_iter):
    """
    Método de Newton para resolver F(x) = 0 usando LU com pivotamento.

    Parâmetros:
    - F: função que recebe x e retorna F(x), vetor em R^m
    - JF: função que recebe x e retorna a matriz Jacobiana JF(x)
    - x0: vetor inicial
    - tol: tolerância relativa para critério de parada
    - max_iter: número máximo de iterações

    Retorna:
    - x: aproximação da solução
    """
    x = np.copy(x0).astype(float)

    for k in range(max_iter):
        xold = np.copy(x)

        Fxk = F(x)           # calcula F(xk)
        JFxk = JF(x)         # calcula JF(xk)

        LU, p = lu_dcmp(JFxk)          # decomposição LU
        c = lu_solve(LU, p, -Fxk)      # resolve JFxk * c = -F(xk)
        x = xold + c                   # nova aproximação

        # Critério de parada (erro relativo)
        norma_c = np.max(np.abs(c))
        norma_x = np.max(np.abs(x))
        if norma_c < tol * max(norma_x, 1e-15):  # evita divisão por 0
            print('Convergência em', k + 1, 'iterações.')
            return x

    print('Número máximo de iterações atingido.')
    return x


## Testando a implementação do Método de Newton

Para os dois testes abaixo, use $tol = 10^{-10}$.

1. Calcule o ponto de mínimo da função $H(x,y) = (x-2)^2 + (y-3)^2$, determinando para tanto o ponto onde o seu gradiente se anula.

In [9]:
def F(x):
    # Gradiente de H(x, y) = (x - 2)^2 + (y - 3)^2
    return np.array([
        2 * (x[0] - 2),
        2 * (x[1] - 3)
    ])

def JF(x):
    # Jacobiana da função F (nabla H)
    return np.array([
        [2, 0],
        [0, 2]
    ])

x0 = np.array([0.0, 0.0])  # chute inicial
tol = 1e-10
max_iter = 20

x = newton_sitemas(F, JF, x0, tol, max_iter)

print("Ponto de mínimo encontrado:", x)


Convergência em 2 iterações.
Ponto de mínimo encontrado: [2. 3.]


Quantas iterações foram necessárias? \\
**Resposta:** 1 iteração.

A solução calculada é exata?         
**Resposta:** Sim, a solução encontrada é exatamente x= [2.0 , 3.0].

Você sabe explicar o resultado?      
**Responda com uma frase:** O método converge em uma única iteração porque a função é quadrática e seu gradiente é linear, com Jacobiana constante.

2. Dada a função $F(x_1,x_2,x_3,x_4) = (4x_1-x_2+x_3-x_1x_4, -x_1+3x_2-2x_3-x_2x_4,x_1-2x_2+3x_3-x_3x_4,x_1^2+x_2^2+x_3^2-1)$, determine a raiz que se
obtém pelo método de Newton tomando $x_0 = (1, 1, 1, 1)$. Lembre-se de usar $tol = 10^{-10}$. Imprima a solução calculada.

In [11]:
# Você sabe o que fazer. Defina as funções F e JF, e use o método de Newton

def F(x):
    x1, x2, x3, x4 = x
    return np.array([
        4*x1 - x2 + x3 - x1*x4,
        -x1 + 3*x2 - 2*x3 - x2*x4,
        x1 - 2*x2 + 3*x3 - x3*x4,
        x1**2 + x2**2 + x3**2 - 1
    ])

def JF(x):
    x1, x2, x3, x4 = x
    return np.array([
        [4 - x4,    -1,      1,    -x1],
        [-1,     3 - x4,   -2,    -x2],
        [1,     -2,     3 - x4,   -x3],
        [2*x1,   2*x2,    2*x3,     0 ]
    ])

x0 = np.array([1.0, 1.0, 1.0, 1.0])
tol = 1e-10
max_iter = 50

x = newton_sitemas(F, JF, x0, tol, max_iter)

print("Solução encontrada:")
print(x)


Convergência em 6 iterações.
Solução encontrada:
[0.         0.70710678 0.70710678 1.        ]


## GPS

Agora você irá usar o Método de Newton para calcular a posição $(x,y,z)$ e o erro de sincronia $T = w/c$ de um receptor de sinais de GPS. Os dados serão extraídos do arquivo *input_gps.txt* no qual estão armazenados números dispostos em $35$ linhas e $6$ colunas. Os dados estão associados a $5$ instantes de medidas e em cada instante foram usadas informações de $7$ de $25$ satélites disponíveis. Os intantes estão espaçados de $15$ em $15$ minutos.

Para cada um destes instantes temos os valores $x_i$, $y_i$, $z_i$ e $w_i$ em metros, $1\le i\le 7$, usados de 7 satélites (os lapsos de tempo aparentes já foram convertidos para metros). Os dados no arquivo *input_gps.txt* estão dispostos da seguinte forma:

> **Primeira coluna:** número identificando o satélite

> **Segunda coluna:** instantes de medida (min) com os valores
- Linhas $1$ a $7$: $\quad \ \ \ \,0$
- Linhas $8$ a $14$: $\quad \ \,15$
- Linhas $15$ a $21$: $\quad 30$
- Linhas $22$ a $28$: $\quad 45$
- Linhas $29$ a $35$: $\quad 60$

> **3ª, 4ª e 5ª colunas:** coordenadas dos $x_i$, $y_i$ e $z_i$ dos satélites, respectivamente, em metros

> **Sexta coluna:** lapsos de tempo aparentes $w_i$, já convertidos em metros

Você irá usar **somente os dados de $7$ satélites correspondentes a um único instante de medida**, calculado como o resto da divisão do seu número USP por $5$, multiplicado por $15$: $$t_m = (NUSP\mod5)*15.$$

Organize o seu programa da seguinte forma:

1. Salve o conteúdo arquivo *input_gps.txt* em um numpy array com o nome input_gps. Extraia do array input_gps a parte correspondente ao seu número USP na forma de uma matriz $7\times 6$ e imprima esta matriz. Isto ajudará a verificar se você pegou os dados corretos, pois na segunda coluna deve haver 7 números iguais correspondentes ao instante de medida.

In [12]:
# Leitura do arquivo input_gps.txt, salvando o conteúdo em input_gps
import gdown

# Only the file ID goes here
file_id = '1NRvx1o3hlQiwVsigVSQ2zJ_MecG78Pbu'
url = f'https://drive.google.com/uc?id={file_id}&confirm=t'

# Download the file
gdown.download(url, 'input_gps.txt', quiet=False)

# Load it with NumPy
import numpy as np
input_gps = np.loadtxt('input_gps.txt')

Downloading...
From: https://drive.google.com/uc?id=1NRvx1o3hlQiwVsigVSQ2zJ_MecG78Pbu&confirm=t
To: /content/input_gps.txt
100%|██████████| 3.96k/3.96k [00:00<00:00, 6.84MB/s]


In [13]:
# Calcular o instante de medida com base no número USP
NUSP = 12610305
tm = (NUSP % 5) * 15  # em minutos

# Filtrar as linhas com instante igual a tm
instante_desejado = tm
dados_instante = input_gps[input_gps[:, 1] == instante_desejado]

# Verificação: deve conter 7 linhas e a segunda coluna deve conter apenas tm
assert dados_instante.shape == (7, 6), "Erro: não foram encontrados 7 satélites para o instante desejado."
assert np.all(dados_instante[:, 1] == tm), "Erro: instante de medida inconsistente."

# Exibir a matriz 7x6 correspondente ao instante
print(f"Matriz de satélites para t = {tm} minutos:")
print(dados_instante)


Matriz de satélites para t = 0 minutos:
[[ 1.00000000e+00  0.00000000e+00 -1.71043262e+07 -5.22851009e+06
   1.98114707e+07  2.34869342e+07]
 [ 2.00000000e+00  0.00000000e+00 -3.77208364e+06 -2.64154682e+07
   1.17640822e+05  2.24307618e+07]
 [ 1.60000000e+01  0.00000000e+00  2.60036705e+06 -2.58041067e+07
   5.85324527e+06  2.12819017e+07]
 [ 2.00000000e+01  0.00000000e+00 -1.30610557e+07 -1.49746105e+07
   1.75417460e+07  2.15423939e+07]
 [ 2.50000000e+01  0.00000000e+00  8.15188814e+06 -1.36415314e+07
   2.16241596e+07  2.12150373e+07]
 [ 1.40000000e+01  0.00000000e+00  2.08870794e+07 -1.61598188e+07
   3.06898076e+06  2.36749103e+07]
 [ 6.00000000e+00  0.00000000e+00  2.26660343e+07 -2.99004344e+06
   1.38428759e+07  2.45993565e+07]]


In [14]:
# Instante das medições
tm = (12610305 % 5) * 15  # => tm = 0
print('tm =', tm)
print()

# Extração dos dados para o instante tm do array input_gps
meus_dados = input_gps[input_gps[:, 1] == tm]

# Impressão (restringindo a precisão para caber na tela)
for row in meus_dados:
    # First two columns as integers
    print(f"{int(row[0]):3d} {int(row[1]):3d} ", end="")

    # Remaining columns with sete casas significativas
    print(" ".join(f"{val:15.6e}" for val in row[2:]))


tm = 0

  1   0   -1.710433e+07   -5.228510e+06    1.981147e+07    2.348693e+07
  2   0   -3.772084e+06   -2.641547e+07    1.176408e+05    2.243076e+07
 16   0    2.600367e+06   -2.580411e+07    5.853245e+06    2.128190e+07
 20   0   -1.306106e+07   -1.497461e+07    1.754175e+07    2.154239e+07
 25   0    8.151888e+06   -1.364153e+07    2.162416e+07    2.121504e+07
 14   0    2.088708e+07   -1.615982e+07    3.068981e+06    2.367491e+07
  6   0    2.266603e+07   -2.990043e+06    1.384288e+07    2.459936e+07


2. Você irá trabalhar somente com as terceira, quarta, quinta e sexta colunas da matriz de dados do item 1. Com estes dados, construa o sistema linear sobredeterminado $(3)$ e **resolva o sistema normal usando a decomposição** $LU$ que você implementou. Armazene e imprima a solução. Ela será usada como condição inicial para o Método de Newton.



In [16]:
#***IMPLEMENTE AQUI A SOLUÇÃO DO SISTEMA NORMAL (a resposta deve ser um array com 4 elementos)

# Extrai apenas as colunas x_i, y_i, z_i, w_i
sat_coords = meus_dados[:, 2:6]

# Satélite de referência (o 7º)
xr, yr, zr, wr = sat_coords[6]

# Construção do sistema linear A * x = b
A = []
b = []

for i in range(6):  # satélites 0 a 5 (primeiros 6)
    xi, yi, zi, wi = sat_coords[i]

    ai = [
        2 * (xr - xi),
        2 * (yr - yi),
        2 * (zr - zi),
        2 * (wr - wi)
    ]
    bi = xr**2 - xi**2 + yr**2 - yi**2 + zr**2 - zi**2 + wr**2 - wi**2

    A.append(ai)
    b.append(bi)

A = np.array(A)
b = np.array(b)

# Sistema normal: Aᵗ A x = Aᵗ b
AtA = A.T @ A
Atb = A.T @ b

# Resolver com LU
LU, p = lu_dcmp(AtA)
x_normal = lu_solve(LU, p, Atb)

# Impressão da solução
print('Solução do sistema normal:', "".join(f"{val:15.6e}" for val in x_normal))


Solução do sistema normal:    3.020028e+05  -4.326742e+06   3.824092e+06   4.303395e+07


3. Com os mesmos dados do item anterior, programe duas funções: uma para $F=\nabla g$ (queremos achar raízes da equação $(2)$) e outra para a matriz Jacobiana $J_F$ de $F$. Resolva o sistema não linear $F(\mathbf{x})=0$ para obter a solução $\bar{\mathbf{x}} = (\bar x, \bar y, \bar z, \bar w)$ da equação $(2)$ usando a implementação do Método de Newton descrita acima. Armazene e imprima a solução, e imprima separadamente o erro de sincronia $T=\bar w / c$. Qual é o erro relativo entre as soluções de $(2)$ e $(3)$?



In [17]:
# Matriz com os dados dos satélites: colunas (x_i, y_i, z_i, w_i)
xyzw = meus_dados[:, 2:6]
c = 299_792_458  # velocidade da luz em m/s

# Função F(x) = gradiente de g(x, y, z, w)
def F(x):
    x0, y0, z0, w0 = x
    Fx = np.zeros(4)
    for xi, yi, zi, wi in xyzw:
        r = (x0 - xi)**2 + (y0 - yi)**2 + (z0 - zi)**2 - (wi - w0)**2
        Fx[0] += 2 * r * 2 * (x0 - xi)
        Fx[1] += 2 * r * 2 * (y0 - yi)
        Fx[2] += 2 * r * 2 * (z0 - zi)
        Fx[3] += 2 * r * (-2) * (wi - w0)
    return Fx

# Jacobiana de F (ou seja, Hessiana de g)
def JF(x):
    x0, y0, z0, w0 = x
    J = np.zeros((4, 4))
    for xi, yi, zi, wi in xyzw:
        dx = x0 - xi
        dy = y0 - yi
        dz = z0 - zi
        dw = wi - w0
        r = dx**2 + dy**2 + dz**2 - dw**2

        J[0, 0] += 4 * dx**2 + 2 * r * 2
        J[1, 1] += 4 * dy**2 + 2 * r * 2
        J[2, 2] += 4 * dz**2 + 2 * r * 2
        J[3, 3] += 4 * dw**2 + 2 * r * 2

        J[0, 3] += -8 * dx * dw
        J[1, 3] += -8 * dy * dw
        J[2, 3] += -8 * dz * dw

    # simetria da hessiana
    J[3, 0] = J[0, 3]
    J[3, 1] = J[1, 3]
    J[3, 2] = J[2, 3]
    return J

# Resolução do sistema F(x) = 0 via método de Newton
x_barra = newton_sitemas(F, JF, x_normal, tol=1e-10, max_iter=50)

# Imprime a solução
print("Solução pelo Método de Newton (x̄, ȳ, z̄, w̄):")
print("".join(f"{val:15.6e}" for val in x_barra))

# Imprime T = w / c
T = x_barra[3] / c
print(f"\nErro de sincronia T = w / c: {T:.12f} segundos")

# Calcula e imprime o erro relativo entre x_barra e x_normal
erro_relativo = np.linalg.norm(x_barra - x_normal, ord=2) / np.linalg.norm(x_barra, ord=2)
print(f"\nErro relativo entre soluções do sistema normal e sistema não linear: {erro_relativo:.12e}")


Convergência em 33 iterações.
Solução pelo Método de Newton (x̄, ȳ, z̄, w̄):
   2.908267e+06  -1.303348e+07   1.153582e+07   2.305752e+07

Erro de sincronia T = w / c: 0.076911614438 segundos

Erro relativo entre soluções do sistema normal e sistema não linear: 8.011658751737e-01


4. Para as soluções de $(2)$ e $(3)$, determine as latitudes, longitudes e elevações (coordenadas geográficas) e localize-as no globo terrestre.

In [18]:
def cartesiano_para_geografico(xyz):
    x, y, z = xyz[:3]  # pega apenas os 3 primeiros (x, y, z)
    R = 6_371_000  # raio médio da Terra em metros

    lon = np.arctan2(y, x)
    r_xy = np.sqrt(x**2 + y**2)
    lat = np.arctan2(z, r_xy)
    alt = np.sqrt(x**2 + y**2 + z**2) - R

    # Converter radianos para graus
    lon_deg = np.degrees(lon)
    lat_deg = np.degrees(lat)

    return lat_deg, lon_deg, alt

# Coordenadas da solução do sistema normal (3)
lat3, lon3, alt3 = cartesiano_para_geografico(x_normal)

# Coordenadas da solução do sistema não linear (2)
lat2, lon2, alt2 = cartesiano_para_geografico(x_barra)

# Impressão
print("Solução do sistema linear (3):")
print(f"Latitude:  {lat3:.6f}°")
print(f"Longitude: {lon3:.6f}°")
print(f"Altitude:  {alt3:.2f} m\n")

print("Solução do sistema não linear (2):")
print(f"Latitude:  {lat2:.6f}°")
print(f"Longitude: {lon2:.6f}°")
print(f"Altitude:  {alt2:.2f} m")


Solução do sistema linear (3):
Latitude:  41.402046°
Longitude: -86.007280°
Altitude:  -588649.02 m

Solução do sistema não linear (2):
Latitude:  40.821977°
Longitude: -77.421197°
Altitude:  11275667.20 m


Latidude, longitude e elevação correspondentes à solução do sistema normal:

**Resposta**
Latitude: 41.402046°
Longitude: -86.007280°
Elevação: -588.649,02 m

Latidude, longitude e elevação correspondentes à solução do sistema não linear:

**Resposta**
Latitude: 40.821977°
Longitude: -77.421197°
Elevação: 11.275.667,20 m

As coordenadas geográficas foram obtidas a partir das soluções cartesianas (x, y, z) dos sistemas (3) e (2). Para isso, apliquei fórmulas conhecidas de conversão de  coordenadas cartesianas para geográficas, assumindo a Terra como uma esfera de raio médio R = 6.371 × 10^6 metros. A latitude foi calculada por arctan2(z, sqrt(x^2 + y^2)), a longitude por arctan2(y, x), e a elevação pela diferença entre a norma do vetor posição  e o raio da Terra. As coordenadas resultantes foram expressas em graus e metros.